src/python/pants/engine/legacy/options_parsing.py

```diff
@@ -0,0 +1,48 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+from pants.engine.rules import RootRule, rule
+from pants.engine.selectors import Select
+from pants.init.options_initializer import OptionsInitializer
+from pants.option.options_bootstrapper import OptionsBootstrapper
+from pants.util.objects import datatype
+
+
+class OptionsParseRequest(datatype(['args', 'env'])):
+  """Represents a request for Options computation."""
+
+  @classmethod
+  def create(cls, args, env):
+    assert isinstance(args, (list, tuple))
+    return cls(
+      tuple(args),
+      tuple(env.items() if isinstance(env, dict) else env)
+    )
+
+
+class Options(datatype(['options', 'build_config'])):
+  """Represents the result of an Options computation."""
+
+
+@rule(Options, [Select(OptionsParseRequest)])
+def parse_options(parse_request):
+  options_bootstrapper = OptionsBootstrapper(
+    env=dict(parse_request.env),
+    args=parse_request.args
+  )
+  options_initializer = OptionsInitializer(options_bootstrapper)
+  options, build_config = options_initializer.setup(init_logging=False, init_subsystems=False)
+  options.freeze()
+  return Options(options, build_config)
+
+
+def create_options_parsing_rules():
+  return [
+    parse_options,
+    RootRule(Options),
```

Comment:
Shouldn't need a RootRule for `Options`, because it isn't passed in as a parameter to a request. Hopefully this will make more sense post #5788.

Reply:
ah, k - for some reason the rule validation logic led me to this, but you're right it doesn't appear to be necessary. removed.

Reply:
`src/python/pants/engine/README.md` was updated in #5759 with some description of `RootRule` if you hadn't seen it yet.